libraries/AP_HAL_ChibiOS/AnalogIn.cpp

```diff
@@ -206,233 +236,389 @@ void AnalogSource::_add_value(float v, float vcc5V)
+                sample_sum[1][j] += *buffer;
+                buffer++;
+            } 
+#endif
         }
     }
-    sample_count += ADC_DMA_BUF_DEPTH;
+#if STM32_ADC_DUAL_MODE
+    if (index == 0) {
+        // also set the sample count for the second ADC for dual mode
+        sample_count[1] += ADC_DMA_BUF_DEPTH;
+    }
+#endif
+    sample_count[index] += ADC_DMA_BUF_DEPTH;
+}
+
+uint8_t AnalogIn::get_adc_index(ADCDriver* adcp)
+{
+    if (adcp == &ADCD1) {
+        return 0;
+    }
+#if defined(HAL_ANALOG2_PINS) && !STM32_ADC_DUAL_MODE
+    if (adcp == &ADCD2) {
+        return 1;
+    }
+#endif
+#if defined(HAL_ANALOG3_PINS)
+    if (adcp == &ADCD3) {
+        return 2;
+    }
+#endif
+    osalDbgAssert(false, "invalid ADC");
+    return 255;
+}
+
+uint8_t AnalogIn::get_num_grp_channels(uint8_t index)
+{
+    switch (index) {
+    case 0:
+        return ADC_GRP1_NUM_CHANNELS;
+#if defined(HAL_ANALOG2_PINS)
+    case 1:
+        return ADC2_GRP1_NUM_CHANNELS;
+#endif
+#if defined(HAL_ANALOG3_PINS)
+    case 2:
+        return ADC3_GRP1_NUM_CHANNELS;
+#endif
+    };
+    osalDbgAssert(false, "invalid adc_index");
+    return 0;
 }
 
 /*
   setup adc peripheral to capture samples with DMA into a buffer
  */
 void AnalogIn::init()
 {
+#if STM32_ADC_DUAL_MODE
+    static_assert(sizeof(uint32_t) == sizeof(adcsample_t), "adcsample_t must be uint32_t");
+#else
     static_assert(sizeof(uint16_t) == sizeof(adcsample_t), "adcsample_t must be uint16_t");
+#endif
+    for (uint8_t i=0; i<HAL_NUM_ANALOG_INPUTS; i++) {
+        setup_adc(i);
+    }
+}
 
-    if (ADC_GRP1_NUM_CHANNELS == 0) {
+void AnalogIn::setup_adc(uint8_t index)
+{
+    uint8_t num_grp_channels = get_num_grp_channels(index);
+    if (num_grp_channels == 0) {
         return;
     }
 
-    samples = (adcsample_t *)hal.util->malloc_type(sizeof(adcsample_t)*ADC_DMA_BUF_DEPTH*ADC_GRP1_NUM_CHANNELS, AP_HAL::Util::MEM_DMA_SAFE);
+    ADCDriver *adcp;
+    switch (index) {
+    case 0:
+        adcp = &ADCD1;
+        break;
+// if we are in dual mode then ADC2 is setup along with ADC1
+// so we don't need to setup ADC2 separately
+#if defined(HAL_ANALOG2_PINS) && !STM32_ADC_DUAL_MODE
+    case 1:
+        adcp = &ADCD2;
+        break;
+#endif
+#if defined(HAL_ANALOG3_PINS)
+    case 2:
+        adcp = &ADCD3;
+        break;
+#endif
+    default:
+        return;
+    };
+
+#if STM32_ADC_DUAL_MODE
+    // In ADC Dual mode we use ADC_SAMPLE_SIZE as 32 bits for ADCs doing shared sampling, then we need to
+    // split the samples into two buffers at read time for other ADCs ChibiOS HAL uses 16 bit sample size
+    // so ADC_SAMPLE_SIZE is actually 16 bits for ADCs even though its set to 32 bits
+    if (index == 0) {
+        // we need to setup the second ADC as well
+        num_grp_channels = num_grp_channels * 2;
+        samples[0] = (adcsample_t *)hal.util->malloc_type(sizeof(uint16_t)*ADC_DMA_BUF_DEPTH*num_grp_channels, AP_HAL::Util::MEM_DMA_SAFE);
```

Comment:
adcsample_t?

Reply:
samples are of type adcsample_t, they are 32bit wide for dual mode, 16bit otherwise. Here we allocate 2x16bit ints